tiledb/api/src/query/read/aggregate/mod.rs

```diff
@@ -0,0 +1,1054 @@
+    Min(String),
+    /// Computes the maximum value of the argument field.
+    Max(String),
+    /// Computes the sum of the argument field.
+    Sum(String),
+    /// Computes
+    Mean(String),
+}
+
+impl AggregateFunction {
+    /// Returns the name of the field which this function applies to.
+    pub fn argument_name(&self) -> Option<&str> {
+        match self {
+            Self::Count => None,
+            Self::NullCount(ref s)
+            | Self::Min(ref s)
+            | Self::Max(ref s)
+            | Self::Sum(ref s)
+            | Self::Mean(ref s) => Some(s.as_ref()),
+        }
+    }
+
+    /// Returns a unique name for this aggregate function.
+    pub fn aggregate_name(&self) -> String {
+        match self {
+            Self::Count => "Count".to_owned(),
+            Self::NullCount(ref s) => format!("NullCount({})", s),
+            Self::Min(ref s) => format!("Min({})", s),
+            Self::Max(ref s) => format!("Max({})", s),
+            Self::Sum(ref s) => format!("Sum({})", s),
+            Self::Mean(ref s) => format!("Mean({})", s),
+        }
+    }
+
+    /// Returns the result type of the aggregate operation.
+    ///
+    /// This is used to determine if the user's requested programmatic data type
+    /// is compatible with the result type of the aggregate function.
+    pub(crate) fn result_type_impl(
+        &self,
+        argument_type: Option<(Datatype, CellValNum)>,
+    ) -> Option<Datatype> {
+        let is_unit = matches!(argument_type.map(|(_, cvn)| cvn), Some(CellValNum::Fixed(nz)) if nz.get() == 1);
+
+        match self {
+            AggregateFunction::Count | AggregateFunction::NullCount(_) => {
+                Some(Datatype::UInt64)
+            }
+            AggregateFunction::Mean(_) => {
+                if is_unit {
+                    Some(Datatype::Float64)
+                } else {
+                    None
+                }
+            }
+            AggregateFunction::Min(_) | AggregateFunction::Max(_) => {
+                if is_unit {
+                    Some(argument_type.unwrap().0)
+                } else {
+                    None
+                }
+            }
+            AggregateFunction::Sum(_) => {
+                if is_unit {
+                    let argument_type = argument_type.unwrap().0;
+                    match argument_type {
+                        Datatype::Int8
+                        | Datatype::Int16
+                        | Datatype::Int32
+                        | Datatype::Int64 => Some(Datatype::Int64),
+                        Datatype::UInt8
+                        | Datatype::UInt16
+                        | Datatype::UInt32
+                        | Datatype::UInt64 => Some(Datatype::UInt64),
+                        Datatype::Float32 | Datatype::Float64 => {
+                            Some(Datatype::Float64)
+                        }
+                        _ => None,
```

Comment:
Is this dictated by core or am I missing a Rust reason for the coercion to these types?

Reply:
I think this is dictated by core. But not in a way that's exposed by the API.

If it is then we should find the code that dictates it and leave a note here.

Reply:
From what I can tell:

The C API reasons about buffer sizes, not about the type of the pointee.  So the output buffer validation code basically does the same thing.  Each aggregate has its own way of calling a function in `output_buffer_validator.cc`.

Then each aggregate has its own way of copying its intermediate result to the final output.  Sum has `sum_type_data`, Mean just always writes double.

All this to say, I think there's a few references we can add here, but not a single function

Reply:
22126ded